File: data/figures/chapter02/figure_2_3/plot_2_3.m
function outputPath = plot_2_3(outputPath)
%PLOT_2_3 Original 1.002 mm coordinate repair, all receiving skin facets.
% Requires only figure_2_3.csv. No solver, archive, or external data dependency.
% Contract: preserve true oblique/side coordinates and equal scales, reveal
% the turning skin closure at C, distinguish thin skin edges from thick beams.
out=fileparts(mfilename('fullpath'));
if nargin<1 || isempty(outputPath),outputPath=fullfile(out,'figure_2_3.png');end
outputPath=char(outputPath);
[outputFolder,~,extension]=fileparts(outputPath);
if ~any(strcmpi(extension,{'.png','.pdf','.svg'}))
    error('plot_2_3:OutputFormat','Output must use PNG, PDF or SVG.');
end
if ~isempty(outputFolder) && ~isfolder(outputFolder),mkdir(outputFolder);end
d=readtable(fullfile(out,'figure_2_3.csv'),'TextType','string');
v=xyz(d(d.entity=="receiving_vertex" & d.receiver=="B",:));
a=xyz(d(d.entity=="beam_node" & d.receiver=="A" & d.state=="original",:));
b=xyz(d(d.entity=="beam_node" & d.receiver=="B" & d.state=="original",:));
c=xyz(d(d.entity=="beam_node" & d.receiver=="C" & d.state=="original",:));
p=xyz(d(d.entity=="projection" & d.receiver=="B",:));
n=cross(v(2,:)-v(1,:),v(3,:)-v(1,:));n=n/norm(n);if dot(p-b,n)<0,n=-n;end
t=c-a;t=t-dot(t,n)*n;t=t/norm(t);s=cross(n,t);basis=[t(:),s(:),n(:)];
assert(norm(basis.'*basis-eye(3),'fro')<1e-12);
tr=@(rows) (xyz(rows)-p)*basis*1000;
q=tr(d);xl=[min(q(:,1))-.7,max(q(:,1))+.9];
yl=[min(q(:,2))-.3,max(q(:,2))+.3];zl=[min(q(:,3))-.7,max(q(:,3))+.7];
blue=[.19 .39 .62];orange=[.78 .38 .12];ink=[.17 .19 .21];grey=[.50 .55 .59];
fig=figure('Visible','off','Color','w','Position',[80 80 1400 1030]);
set(fig,'DefaultAxesFontName','Arial','DefaultTextFontName','Arial');
note(fig,[.045 .950 .91 .042],'Single-node repair at the skin mesh',22,true,ink);
note(fig,[.045 .902 .91 .038], ...
    'A and C are already close to their respective skin facets. Only B is relocated; all three receiving facets are shown.',12,false,ink);
headings={'(a) Oblique view - before repair','(b) Oblique view - after repair', ...
    '(c) Side view - before repair','(d) Side view - after repair'};
positions=[.035 .520 .445 .33;.525 .520 .445 .33;.035 .238 .445 .22;.525 .238 .445 .22];
for panel=1:4
    before=mod(panel,2)==1;side=panel>2;
    state="original";if ~before,state="final";end
    ax=axes(fig,'Position',positions(panel,:));hold(ax,'on');
    for entity=["context_vertex","receiving_vertex"]
        for eid=unique(d.element_id(d.entity==entity)).'
            tri=d(d.entity==entity & d.element_id==eid,:);verts=tr(tri);
            if side,verts=verts(:,[1 3]);end
            edge=grey;face=[.86 .88 .90];alpha=.10;width=.75;
            if entity=="receiving_vertex",edge=blue;face=blue;alpha=.23;width=1.35;end
            patch(ax,'Vertices',verts,'Faces',[1 2 3],'FaceColor',face,'FaceAlpha',alpha,'EdgeColor',edge,'LineWidth',width);
        end
    end
    for eid=[936457 936828]
        verts=tr(d(d.entity=="beam_endpoint" & d.element_id==eid & d.state==state,:));
        if side
            plot(ax,verts(:,1),verts(:,3),'-','Color',orange,'LineWidth',3.4);
        else
            plot3(ax,verts(:,1),verts(:,2),verts(:,3),'-','Color',orange,'LineWidth',3.4);
        end
    end
    for name=["A","B","C"]
        node=tr(d(d.entity=="beam_node" & d.receiver==name & d.state==state,:));
        fill=ink;edge=ink;sz=8;
        if name=="B",fill='w';edge=orange;sz=10;if ~before,fill=orange;end,end
        label=char(name);if name=="B" && ~before,label=['B' char(8242)];end
        offset=[-.25 0 .28];
        if name=="B",offset=[-.12 0 -.40];elseif name=="C",offset=[.32 0 -.05];end
        if name=="C" && side,offset(1)=.62;end
        if name=="B" && ~before && ~side,offset(3)=-.70;end
        if side
            plot(ax,node(1),node(3),'o','MarkerFaceColor',fill,'MarkerEdgeColor',edge,'MarkerSize',sz,'LineWidth',1.6);
            text(ax,node(1)+offset(1),node(3)+offset(3),label,'Color',ink,'FontSize',14,'FontWeight','bold');
        else
            plot3(ax,node(1),node(2),node(3),'o','MarkerFaceColor',fill,'MarkerEdgeColor',edge,'MarkerSize',sz,'LineWidth',1.6);
            text(ax,node(1)+offset(1),node(2),node(3)+offset(3),label,'Color',ink,'FontSize',14,'FontWeight','bold');
        end
    end
    if side
        if before
            plot(ax,0,0,'x','Color',blue,'MarkerSize',8,'LineWidth',1.3);
            text(ax,-.35,.25,'Projection of B','Color',blue,'FontSize',10);
        end
        % Actual closure facet projects downward from the upper skin at C.
        closure=tr(d(d.entity=="receiving_vertex" & d.receiver=="C",:));
        k=mean(closure,1);
        text(ax,k(1)+.9,k(3)+.45,'Skin closure','Color',blue,'FontSize',11);
        upper=tr(d(d.entity=="receiving_vertex" & d.receiver=="A",:));
        k=mean(upper,1);
        text(ax,k(1)-2.0,k(3)+.42,'Upper skin','Color',blue,'FontSize',11);
        axis(ax,'equal');xlim(ax,xl);ylim(ax,zl);
        sx=xl(1)+.25;sz=zl(1)+.25;
        plot(ax,[sx sx+1],[sz sz],'-','Color',ink,'LineWidth',1.3);
        text(ax,sx+.5,sz-.25,'1 mm','Color',ink,'FontSize',10,'HorizontalAlignment','center');
    else
        axis(ax,'equal');xlim(ax,xl);ylim(ax,yl);zlim(ax,zl);
        view(ax,[10 28]);camproj(ax,'orthographic');
    end
    set(ax,'Visible','off','Clipping','off');
    title(ax,headings{panel},'Visible','on','FontSize',14,'FontWeight','normal','Color',ink);
end
annotation(fig,'line',[.065 .105],[.180 .180],'Color',grey,'LineWidth',.9);
note(fig,[.111 .163 .27 .034],'Skin-element edges',11,false,ink);
annotation(fig,'line',[.380 .420],[.180 .180],'Color',orange,'LineWidth',3.4);
note(fig,[.426 .163 .27 .034],'Beam centre-lines',11,false,ink);
annotation(fig,'rectangle',[.725 .170 .025 .021],'FaceColor',blue*.25+.75,'EdgeColor',blue,'LineWidth',1.1);
note(fig,[.757 .163 .23 .034],'Receiving skin elements',11,false,ink);
note(fig,[.045 .109 .91 .036], ...
    'B coordinate correction: 1.002 mm. A and C retain their coordinates and are MPC-coupled to the skin.',12,true,ink);
note(fig,[.045 .061 .91 .034], ...
    'C is close to the turning closure facet, not far from the skin. Equal length scales; no added beam or load-induced deformation.',10,false,ink);
note(fig,[.045 .023 .91 .030], ...
    'Source data: accompanying CSV. All shown skin triangles form one edge-connected patch; blue facets receive A, B and C.',9,false,grey);
set(findall(fig,'-property','Interpreter'),'Interpreter','none');
set(fig,'PaperPositionMode','auto');
switch lower(extension)
    case '.png',print(fig,outputPath,'-dpng','-r220');
    case '.pdf'
        set(fig,'PaperUnits','inches');paper=get(fig,'PaperPosition');
        set(fig,'PaperSize',paper(3:4),'PaperPosition',[0 0 paper(3:4)]);
        print(fig,outputPath,'-dpdf','-r220');
    case '.svg',print(fig,outputPath,'-dsvg','-r220');
end
close(fig);
fprintf('Rendered 3 receiving facets and an edge-connected surrounding skin patch from CSV only.\n');
end

function q=xyz(rows)
q=[rows.x_m rows.y_m rows.z_m];
end
function note(fig,pos,message,size,bold,color)
weight='normal';if bold,weight='bold';end
annotation(fig,'textbox',pos,'String',message,'FontName','Arial','FontSize',size, ...
    'FontWeight',weight,'Color',color,'EdgeColor','none','Interpreter','none');
end

File: data/figures/chapter02/figure_2_3/plot_2_3.py
"""Reproduce the corrected microscopic single-node example from its CSV only.

Install NumPy and Matplotlib, then run: python plot_2_3.py
Use --output PATH for PNG, PDF or SVG; default: figure_2_3_python.png.
No solver, external archive, or extra data
source is used. Skin facets are thin blue/grey edges; beams are thick orange.
The skin closure at C is retained, rather than filtered out as non-coplanar.
"""
import argparse
import csv
from pathlib import Path
import numpy as np
import matplotlib
matplotlib.use("Agg")
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d.art3d import Poly3DCollection

OUT=Path(__file__).resolve().parent
BLUE="#31639e";ORANGE="#c7611f";INK="#2c3035";GREY="#a4adb5"
plt.rcParams.update({"font.family":"DejaVu Sans","font.size":11,"text.color":INK})


def main(output):
    with (OUT/"figure_2_3.csv").open(newline="",encoding="utf-8") as f:rows=list(csv.DictReader(f))
    def pick(**conditions):return [r for r in rows if all(r[k]==v for k,v in conditions.items())]
    def xyz(data):return np.array([[float(r[k]) for k in ["x_m","y_m","z_m"]] for r in data])
    def point(role,state="original"):return xyz(pick(entity="beam_node",receiver=role,state=state))[0]
    p=xyz(pick(entity="projection",receiver="B"))[0]
    v=xyz(pick(entity="receiving_vertex",receiver="B"))
    n=np.cross(v[1]-v[0],v[2]-v[0]);n/=np.linalg.norm(n)
    if np.dot(point("B","final")-point("B"),n)<0:n=-n
    t=point("C")-point("A");t-=np.dot(t,n)*n;t/=np.linalg.norm(t)
    basis=np.column_stack([t,np.cross(n,t),n]);assert np.allclose(basis.T@basis,np.eye(3),atol=1e-12)
    local=lambda data:(xyz(data)-p)@basis*1e3
    faces=[]
    for entity in ["context_vertex","receiving_vertex"]:
        for eid in sorted({r["element_id"] for r in pick(entity=entity)},key=int):
            face=sorted(pick(entity=entity,element_id=eid),key=lambda r:int(r["vertex_index"]))
            faces.append((local(face),face[0]["receiver"] if entity=="receiving_vertex" else ""))
    old=np.array([(point(r)-p)@basis*1e3 for r in "ABC"])
    final=np.array([(point(r,"final")-p)@basis*1e3 for r in "ABC"])
    focus=local(rows)
    margins=[(.7,.9),(.3,.3),(.7,.7)]
    limits=[(focus[:,i].min()-margins[i][0],focus[:,i].max()+margins[i][1]) for i in range(3)]
    fig=plt.figure(figsize=(14,10))
    fig.text(.055,.95,"Single-node repair within the surrounding skin mesh",fontsize=20,weight="bold")
    fig.text(.055,.91,"Only B moves. A and C retain their coordinates and couple to their own skin facets.",fontsize=12)
    for col,state in enumerate(["original","final"]):
        q=old if state=="original" else final
        ax=fig.add_axes([.04+.5*col,.53,.42,.32],projection="3d",computed_zorder=False)
        for vertices,role in faces:
            ax.add_collection3d(Poly3DCollection([vertices],facecolor=BLUE if role else "#e5e8eb",
                edgecolor=BLUE if role else GREY,linewidth=1.1 if role else .6,
                alpha=.24 if role=="B" else (.12 if role else .06),zorder=1))
        ax.plot(*q.T,color=ORANGE,lw=3.5,zorder=3)
        for i,role in enumerate("ABC"):
            ax.scatter(*q[i],s=60,c=ORANGE if role=="B" and state=="final" else ("white" if role=="B" else INK),
                edgecolor=ORANGE if role=="B" else "white",depthshade=False,zorder=4)
            ax.text(q[i,0]-.16,q[i,1],q[i,2]+(.28 if role!="B" else -.45),
                role+("'" if role=="B" and state=="final" else ""),fontsize=12,weight="bold",zorder=5)
        ax.set_xlim(limits[0]);ax.set_ylim(limits[1]);ax.set_zlim(limits[2]);ax.set_proj_type("ortho")
        ax.set_box_aspect([hi-lo for lo,hi in limits],zoom=1.35);ax.view_init(elev=28,azim=-80);ax.set_axis_off()
        ax.set_title("(a) Oblique view: original" if col==0 else "(b) Oblique view: repaired",fontsize=13,pad=10)
        ax=fig.add_axes([.06+.5*col,.18,.385,.27])
        for vertices,role in faces:
            v=vertices[:,[0,2]];v=np.vstack([v,v[0]])
            ax.plot(*v.T,color=BLUE if role else GREY,lw=1.3 if role else .6,zorder=1)
        ax.plot(q[:,0],q[:,2],color=ORANGE,lw=3.5,zorder=3)
        for i,role in enumerate("ABC"):
            ax.scatter(q[i,0],q[i,2],s=60,facecolor=ORANGE if role=="B" and state=="final" else ("white" if role=="B" else INK),
                edgecolor=ORANGE if role=="B" else "white",zorder=4)
            dx=-.42 if role=="B" else (-.2 if role=="A" else .12)
            ax.text(q[i,0]+dx,q[i,2]-.28,role+("'" if role=="B" and state=="final" else ""),fontsize=12,weight="bold",zorder=5)
        # An actual orthographic side view; equal in-plane and normal scales.
        ax.set_aspect("equal");ax.set_xlim(limits[0]);ax.set_ylim(limits[2]);ax.axis("off")
        ax.set_title("(c) Side view: original" if col==0 else "(d) Side view: repaired",fontsize=13,pad=12)
        x=limits[0][0]+.15;y=limits[2][0]+.15
        ax.plot([x,x+1],[y,y],color=INK,lw=1.3);ax.text(x+.5,y-.27,"1 mm",ha="center",fontsize=10)
    correction=np.linalg.norm(point("B","final")-point("B"))*1e3
    assert np.linalg.norm(point("A","final")-point("A"))==0
    assert np.linalg.norm(point("C","final")-point("C"))==0
    fig.text(.055,.13,f"B coordinate correction: {correction:.3f} mm. C is close to the turning skin closure, not far from the entire skin.",fontsize=12,weight="bold")
    fig.text(.055,.086,"Blue: receiving skin facets for A, B and C     |     Grey: neighbouring skin-element edges     |     Orange: beam centre-lines",fontsize=11)
    fig.text(.055,.044,"Coordinate repair before analysis; not load-induced deformation. True local geometry, with no vertical exaggeration or added beam.",fontsize=10)
    output.parent.mkdir(parents=True,exist_ok=True)
    fig.savefig(output,dpi=220);plt.close(fig)
    print(f"Rendered {output.name} from figure_2_3.csv only.")


if __name__=="__main__":
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--output",type=Path,default=OUT/"figure_2_3_python.png",
                        help="Output PNG, PDF or SVG path.")
    main(parser.parse_args().output)
